### PR TITLE
state-utils: Extract keyedReducer and TypeScriptify it

### PR DESCRIPTION
**Pull request by sarayourfriend**
#### Changes proposed in this Pull Request

* Extract keyed-reducer into a package and type it

#### Testing instructions

* Unit tests pass

### CODE REVIEW

File: packages/state-utils/src/action-types.ts [hunk 1]
```diff
@@ -0,0 +1,2 @@
+export const DESERIALIZE = 'DESERIALIZE' as const;
+export const SERIALIZE = 'SERIALIZE' as const;
```
Comment (sirreal) [inline packages/state-utils/src/action-types.ts:2]: Const strings don't gain anythin with a const assertion. 

```suggestion
export const DESERIALIZE = 'DESERIALIZE';
export const SERIALIZE = 'SERIALIZE';
```

But it _looks_ like these strings aren't exported. Is that right? If so, let's use literal strings directly and remove this file.

**Review by sirreal — COMMENTED** [review]
I tried this a while ago in case it's helpful: https://github.com/Automattic/wp-calypso/pull/33801

I think we should completely get rid of lodash `get` and keypath. I don't think complex paths are used and in practice we just want to get items at the top level, often an object of IDs: `{ posts: { [ postId ]: postObject } }`. we actually just want to access a top-level property. That was a big source of complexity and makes the types opaque (based on my recollection of previous efforts).

**Comment by sarayourfriend** [comment]
@sirreal Originally I tried to get rid of keypath's complexity but actually we do use some complex key paths: https://github.com/automattic/wp-calypso/blob/a47e6f94b01786dfdaa98fe9a235212a979bf7e8/client/state/data-layer/wpcom-http/utils.js#L94

@dmsnell I noticed that you added that keyed reducer in https://github.com/Automattic/wp-calypso/pull/17263. Do you know if it would be possible to refactor it to no longer use a complex key and instead just use a single key? That would be one step towards allowing us to type keyed reducer.

**Comment by dmsnell** [comment]
> Do you know if it would be possible to refactor it to no longer use a complex key and instead just use a single key? That would be one step towards allowing us to type keyed reducer.

it's not really complex if I understand you; just a nested property on an action. it's possible now with template literal types to do that, though what's the trouble with that part of the type signature? what would you ideally want it to be?

**Comment by sarayourfriend** [comment]
Sorry, @dmsnell by complex I just meant a compound key rather than a single, non-nested property key. Could these reducers be keyed on a single, non-nested key? That would allow us to remove the usage of `get` and it would greatly simplify the types. As of right now I'm not sure it's possible to type this without doing something similar to how lodash's `get` is typed and creating a specific overload for nested properties (and then an overload per depth). We can certainly do that no problem but if there's an opportunity to simplify the utility here then I would prefer that. If not that's okay too, no worries, was just curious if you could give any insight.

**Comment by jsnajdr** [comment]
Would it help if the nested property syntax supported only `[ 'a', 'b', 'c' ]` and didn't support `'a.b.c'`?

I vaguely remember trying to simplify `keyedReducer` this way maybe a year ago. But I failed because there were usages of nested keys, and they were reasonable and hard to refactor.

**Comment by jsnajdr** [comment]
Another idea: the `keyPath` is used to extract the key from the `action` object and nothing else. It could also be a function:
```js
// pass a `getKey` function
keyedReducer( action => action.meta?.dataLayer?.requestKey, reducer );

// or an old-fashioned single-property name as a shorthand
keyedReducer( 'siteId', reducer );
```

Would that make `keyedReducer` easier to type? For functions, the key type can be inferred from the action type and the `getKey` function. For strings, we can use `ActionType[ getKey ]`.

**Comment by sarayourfriend** [comment]
@jsnajdr Both of those are options I considered but I would prefer to simplify it completely rather than complicate it.

Taking the array approach, we'd define an overload per array length, just like `get` itself does. That way we can accept the function like so:

```ts
function keyedReducer<
	TState,
	TAction extends Action = AnyAction,
	TKey1 extends string,
	TKey2 extends string,
	TKey3 extends string
>(
	keyPath: [TKey1, TKey2, TKey3],
	reducer: Reducer<TState, TAction>,
): Reducer<{ [TKey1]: { [TKey2]: { [TKey3]: TState } } }, TAction>
```

Or something along those lines.

I don't think that the function approach would work because we wouldn't be able to effectively type the return.

```ts
function keyedReducer<
	TState,
	TAction extends Action = AnyAction,
>(
	getKey: (action: TAction) => TState,
	reducer: Reducer<TState, TAction>,
): // what goes here?
```

In short, I believe the array approach is probably what we'll end up having to compromise with, but if it's possible to use a single key property then I would prefer that over having to maintain (at least) three overloads.

**Comment by jsnajdr** [comment]
@saramarcondes I think there is a misunderstanding about the shape of the state that `keyedReducer` produces. The reducer's `TState` doesn't depend on the type of `TKey`. And with a nested `[ 'a', 'b', 'c' ]` key, there is no triple-nested structure in the state. Example:

```js
const r = keyedReducer( 'a.b.c', ( state, action ) => action.value );

const state1 = r( undefined, {
  type: 'SET',
  a: { b: { c: 'apple' } },
  value: 'ios'
} );
expect( state1 ).toEqual( {
  apple: 'ios'
} );

const state2 = r( state1, {
  type: 'SET',
  a: { b: { c: 'google' } },
  value: 'android'
} );
expect( state2 ).toEqual( {
  apple: 'ios',
  google: 'android' 
} );
```

The `a.b.c` structure appears only on the action, nowhere else. The types would be more like:
```js
function keyedReducer<TState, TAction>(
  key: string,
  reducer: Reducer<TState, TAction>
) : Reducer< { [ TAction[ key ] ]: TState }, TAction >;

function keyedReducer<TState, TAction, TKey>(
  getKey: ( TAction ) => TKey,
  reducer: Reducer<TState, TAction>
) : Reducer< { [ TKey ]: TState }, TAction >;
```

There's also special case where `action[ key ]` or `getKey( action )` can be `undefined` instead of `TKey`. The keyed reducer must accept actions of this shape, and it just returns `state` unchanged for them.

**Comment by sarayourfriend** [comment]
Ah, indeed I misunderstood! Thanks for the clarification. Back to the drawing board for this one, maybe the function approach is really the best here so that the action type can be simplified 🤔 This will take some tinkering. Thanks again for the clarification.

**Comment by sarayourfriend** [comment]
@jsnajdr I figured _something_ out that works but it might be super super cumbersome and awful. I might be too far into the weeds on this one and definitely need an external perspective at this point, if you have any time to take another look.

**Comment by dmsnell** [comment]
@beaucollins shared this link with me - [typing `get`'s property access](https://www.typescriptlang.org/play?ts=4.1.3#code/FDAuE8AcFMAIAUBOB7GiLwIagBYBUpoAePAGgQD5YBeBWaAD1GgDsATAZ1g9EQEsWAc2CwA-HUbN2XAAYASAN4CAZtEQIUaCAF8AdIpVrYAJWg9tMkbGvi89Jq06wFAbQDSsARtRqIAXQAuTxZVdQA1bStraLFvLXAsXAIYIjDyUx4KKJjYIJZoADc1bKD4eyknHn4hbJjbcscuVw8veEDg0NgI2pybLp7evMLi3ushosQRcbUAbhBgAHoF2AABUA4AWj5BFmREaGBlAFcWAGNQPmQWWEFoUBJyMslG7l4BQQoAClx3oLJYSDYHClACUpU0vgSQOSxH+8CoChAkRApyuPFgymQyBozmyACNMIggojRrACQAvYkDHIAG2QNKCLgA5D8OEzyEy+GyOSwjusmX5YJguKiWDxqZEYpFtHNRejGJgALaQGlwWi3UCfTHIDkExC6ClMkEzIA) which does it without any tricks - just takes advantage of the template literal types

the downside to this approach is that I don't think Calypso's actions are going to be typed enough to constrain that argument to `keyedReducer` - that is, about all we can do is say it requires a `string` _or_ magically type it so it requires a dot-separated string path. this is because last time I looked we don't have a singular type for our actions which encompass the properties available, even the meta properties available

**Comment by sarayourfriend** [comment]
@dmsnell I think taking the approach that @jsnajdr suggested, I've achieved the ability to have an "optionally strongly typed" `keyPath` getter function that is flexible enough to run over less specified types.

For example, in the case I linked above, we'd just do a simple refactor of it to `a => a.meta.dataLayer.requestKey` and this _should_ work 🤞🏼

**Comment by dmsnell** [comment]
the only thing I'd guard against there @saramarcondes is making sure we aren't contorting our code simply to appease the type checker. if we move from a string to a function will that enhance our safety? no actually because we can crash in the function vs. the `keyedReducer` which is safe with a string. I don't think the function will guarantee any more than a string would either that it's picking a valid nested property.

just like to ask myself what we gain by adding more types when we start changing our code to avoid type holes.

**Comment by sarayourfriend** [comment]
@dmsnell If you check out the code, the function is safe in the refactor as it's wrapped in a try-catch, which is essentially what `get` was doing before. I don't think this is actually a case where we're changing the code _too_ much to appease the type checker and in fact, when we get around to actually writing fully typed keyed reducers we'll get those type guarantees completely rather than what we have now, which is absolutely no guarantee whatsoever.

Additionally, the 99% usecase is covered under the types, of a simple string key and is strongly typed.

**Comment by jsnajdr** [comment]
Thanks @saramarcondes for the update. I'll need some time to digest this -- the types are quite complicated 🙂

File: packages/state-utils/src/keyed-reducer/index.ts [hunk 2]
```diff
@@ -98,24 +123,33 @@ export const keyedReducer = ( keyPath, reducer ) => {
 					}
 					return result;
 				},
-				undefined
+				undefined as SerializationResult< TState > | undefined
```
Comment (jsnajdr) [inline packages/state-utils/src/keyed-reducer/index.ts]: The result of serialization of `TState` doesn't need to be of type `TState`. It's a generic JSON-serializable value that can be then saved or loaded from local storage. What's the best TypeScript type to describe such a JSON value that usually comes from external sources like local storage or REST response? `JSON.parse` has a return type of `any`. Is there anything better?

A typical usage where the serialization/deserialization is not trivial identity is serializing class instances:
```js
const reducer: Reducer< QueryManager >( state: QueryManager = new QueryManager(), action ) => {
  switch ( action.type ) {
    case 'SERIALIZE':
      return state.toJSON();
    case 'DESERIALIZE':
      return QueryManager.fromJSON( state );
    case 'ADD_RESULTS_PAGE':
      return state.addResultsPage( action.results );
  }
} );
```

The serialization and deserialization with reducers introduces a lot of trouble into our types, because it reuses the reducer for three distinct purposes, where the types are different:
- `reducer( serialized, { type: 'DESERIALIZE' } )` receives value of type `JSON` and returns `TState`
- `reducer( state, { type: 'SERIALIZE' } )` receives `TState` and returns `JSON | undefined`
- `reducer( state, { type: 'FOO' } )` is the actual Redux usage, receiving and returning `TState`. And Redux `TState` is never `undefined`, unlike the result of serialization

Maybe making the `SERIALIZE` and `DESERIALIZE` action quack like Redux action although they really aren't Redux was a bad idea that's not compatible with neat types.

A better reimplementation of persistence would be additional methods on the reducer:
```js
const reducer: Reducer< TState > = ( state, action ) => {
  // pure Redux stuff only
};
reducer.serialize = ( state: TState ) => {
  return ser( state ) as JSON;
};
reducer.deserialize = ( json: JSON ) => {
  return deser( state ) as TState;
}
```

Refactoring the current codebase this way shouldn't be hard and would make types much easier. @sgomes what do you think?

Comment (sgomes) [reply packages/state-utils/src/keyed-reducer/index.ts]: In my experience with serialisation code, I'm reluctant to accept any sentence that includes the words "shouldn't be hard", but the refactor sounds like an excellent idea! 👍 It's much more explicit about intention and result, and it could probably help us get rid of some confusing code.

TypeScript is great for finding inconsistencies like this; if something's hard to type, you may want to rethink the approach.

Comment (sarayourfriend) [reply packages/state-utils/src/keyed-reducer/index.ts]: What would be the best way to implement that new API, @jsnajdr? A new function, `withPersistence` that accepts a reducer function, a serialize function and a deserialize function?

```ts
type SerializableReducer< TState, TAction extends Action = AnyAction > = {
    ( state: TState, action: TAction ) => TState;
    serialize?: ( state: TState ) => JSON;
    deserialize?: ( json: JSON ) => TState;
}

function withPersistence < TState, TAction extends Action = AnyAction >(
    reducer: Reducer< TState, TAction >,
    serialize: ( state: TState ) => JSON,
    deserialize: ( json: JSON ) => TState,
): SerializableReducer< TState, TAction > {
    reducer.serialize = serialize;
    reducer.deserialize = deserialize;
    return reducer;
}
```

And then all reducers that previously declared serialization strategies should be wrapped with the new strategy... the problem, however, is that we can currently dispatch a single action handled by many reducers. With the new approach we'd have to traverse the reducer tree and call serialize on all of them... I'm not seeing how this would be easy, but I'm not super familiar with how all this works. Do you have any more details that might help @jsnajdr or @sgomes?

File: packages/state-utils/src/keyed-reducer/index.ts [hunk 3]
```diff
@@ -57,22 +73,28 @@ import { SerializationResult } from 'calypso/state/serialization-result';
...
+ * @returns super-reducer applying reducer over map of keyed items
  */
-export const keyedReducer = ( keyPath, reducer ) => {
+const keyedReducer = <
+	TKey extends string | number,
+	TState,
+	TAction extends Action = AnyAction,
+	TKeyedAction = Record< string, unknown >
+>(
+	keyPath: keyof KeyedReducerAction< TAction, TKeyedAction > | ( ( action: TAction ) => TKey ),
```
Comment (dmsnell) [inline packages/state-utils/src/keyed-reducer/index.ts]: noting that this introduction of a function here seems to indicate a much more powerful key generator. looks like we can choose a different key based on which action comes in, or even which values are in the actions.

this isn't wrong, but it does suggest making more complicated `keyedReducer` instances when this function was supposed to be fairly straightforward/low-power

maybe it would be helpful to point it out and warn about that, that the purpose isn't to build keys based on dynamic properties, but rather it's a workaround to not typing the dot-separate nested paths or use `string`

File: packages/state-utils/src/keyed-reducer/index.ts [hunk 4]
```diff
@@ -57,22 +73,28 @@ import { SerializationResult } from 'calypso/state/serialization-result';
...
+ * @param keyPath lodash-style path to the key in action referencing item in state map
+ * @param reducer applied to referenced item in state map
+ * @returns super-reducer applying reducer over map of keyed items
  */
-export const keyedReducer = ( keyPath, reducer ) => {
+const keyedReducer = <
+	TKey extends string | number,
+	TState,
+	TAction extends Action = AnyAction,
+	TKeyedAction = Record< string, unknown >
```
Comment (dmsnell) [inline packages/state-utils/src/keyed-reducer/index.ts]: `string` seems too restrictive here. I don't remember any constraint that the key must be a string. possibly `unknown` would be a fitting type for the record keys

**Comment by sarayourfriend** [comment]
@dmsnell Thanks so much for your input. The example you shared with me privately is pasted here for other folks to take a look.

https://www.typescriptlang.org/play?#code/C4TwDgpgBAggxsAlgewHZQLwCgq6gHygG9RIAuKAckVTgCcIBbCVYSgXxz0JPAgsoATCPSYs2nPAWKl+VAM4Q6iAIYAbRAC8IHLrh6yBwxcvVad7ANxYssqACUIggK5wlAHgDKwFcAgAaAD5MKAAKeR8-Cm9fAKgVBBRUCngkNABKTGCYv2ssODQIqALnViUKRxc3OndUZ0YAIyVgjDCI2JCABn94xIys4j0oeQB3RGA4AAswhLTUADpZTKIhqTgVRSoaUWZWSjJVqTwGYGc6dHa-KABqKABGayO8Q9x1zaERBl22A6enk7OF0i0AAtPdHn8XsUNtBKCZVBptPsoVIAedhsDFshvMpUABzULpPJ-KDCABmKmcamAvxJxwgp3RlwgENwnE4tj4UAA0hAQE5Kq4PAAZZDIADWzjAUAgAA8-KhBPIMbi8dI6o0lNJ5CBGsg1D0chAWg4nEKao4CnRBO5RRKpYbgYFAtYCqgiuK+QKzdUQraxZLpXKFUqVTQ1YQNU06NrdQ19Y7YoFQmoA1KKBFVT1xkxBdUKj6PEbAukKLz+YI8yK02BE34WoEhuFgSEiOx4srLchrf77bWoMWerMkpkMMEViS0eg20MOW6PV7BABhZClPwx1qeitVuihSglPY9A-rolYIA

For the rest of the people watching this PR. I was VASTLY over-complicating the issue. Looking forward to y'all's input on this one.

File: packages/state-utils/src/keyed-reducer/index.ts (hunk 2 above)
Comment (jsnajdr) [reply packages/state-utils/src/keyed-reducer/index.ts]: > With the new approach we'd have to traverse the reducer tree and call serialize on all of them...

@saramarcondes All we need to make it work seamlessly is to patch the `combineReducers` function to also combine the children's `{de}serialize` methods. Example of how it would be used in the end:
```js
// reducer with no persistence
function user( state, action ) {
  switch ( action.type ) { ... }
}

// reducer with persistence
function posts( state, action ) {
  switch ( action.type ) { ... }
}

posts.deserialize = ( json: any ): TState => {
  if ( state === undefined ) {
    return getInitialState( posts );
  }
  if ( ! validateWithSchema( json, postsSchema ) ) {
    return getInitialState( posts );
  }
  return json as TState;
}

posts.serialize = ( state: TState ): any => {
  return state;
}

// the combined app reducer
const appReducer = combineReducers( {
  user,
  posts,
} );

// the combined app reducer can serialize/deserialize, too
const booState = appReducer.deserialize( jsonFromStorage );
storage.store( 'state', appReducer.serialize( state );
```

The `user` reducer doesn't define any persistence. That means that its state should not be stored at all. In other words, persistence is opt-in. `combinedReducer.serialize` should return an object that doesn't have an `user` key. And `combinedReducer.deserialize` should return the `user` reducer's initial state.

The `posts` reducer defines a de-serialization that validates the JSON with schema. That code can be extracted into a new version of the `withSchemaValidation` helper. Other than that, the serialization/deserialization is an identity, which stores the state verbatim without any transforms.

**Comment by jsnajdr** [comment]
> I was VASTLY over-complicating the issue.

@saramarcondes If I understand the changes correctly, you mainly removed the complicated types for the actions, didn't you? That's what I also planned to suggest after I played with the Redux types over the weekend 🙂

`keyedReducer` only requires that the action is of type `AnyAction`. That type defines the `action.type` field and the index signature that makes the `action[ keyPath]` property access valid. It doesn't need to define anything else.

Do you even need to define the `KeyedReducerAction`? What if the nested reducer and the return value had just the `Reducer<TState>` type, letting `TAction` default to `AnyAction`?

**Comment by sarayourfriend** [comment]
> Do you even need to define the KeyedReducerAction? What if the nested reducer and the return value had just the Reducer<TState> type, letting TAction default to AnyAction?

If you do this, when we try to use the reducer (for example to pass in `@@calypso/INIT`) we'll get the following error:

```
Argument of type '{ type: string; }' is not assignable to parameter of type 'TAction'.
  '{ type: string; }' is assignable to the constraint of type 'TAction', but 'TAction' could be instantiated with a different subtype of constraint 'Action<any>'
```

This is because `AnyAction` is just a sensible default, not an actual enforced shape.

I don't think the complexity in the action type we've got right now is that bad, consumers can just ignore the extra actions if they want or implement them if they'd like to, but this way they're documented in the type.

The main improvement from my recent changes was to remove the complex `keyPath` type and the enforcement that `keyPath` actually matched the potential actions. This isn't necessary according to @dmsnell.

**Comment by beaucollins** [comment]
Kind of a drive-by at this point, but I've come to find that any time I want deserialization _and_ type safety (hi JSON), I have to reach for a parser. `io-ts`'s Decoder is the pattern I use, but not sure if it fits in with what you all are doing here.

https://github.com/gcanti/io-ts/blob/master/Decoder.md
